parser/prism/Translator.cc

```diff
@@ -1086,7 +1086,51 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node, bool preserveCon
 
             auto arg = translate(definedNode->value);
 
-            return make_unique<parser::Defined>(location.join(arg->loc), move(arg));
+            if (!directlyDesugar) {
+                return make_unique<parser::Defined>(location.join(arg->loc), move(arg));
+            }
+
+            ENFORCE(arg != nullptr);
+            ENFORCE(arg->hasDesugaredExpr());
+
+            // Desugar to `::Magic.defined_instance_var(ivar)` or `::Magic.defined_class_var(cvar)`
+            if (auto *ivar = parser::NodeWithExpr::cast_node<parser::IVar>(arg.get())) {
+                auto sym = MK::Symbol(arg->loc, ivar->name);
+                auto expr = MK::Send1(arg->loc, MK::Magic(arg->loc), core::Names::definedInstanceVar(),
+                                      location.copyWithZeroLength(), move(sym));
+                return make_node_with_expr<parser::Defined>(move(expr), location.join(arg->loc), move(arg));
+            }
+
+            if (auto *cvar = parser::NodeWithExpr::cast_node<parser::CVar>(arg.get())) {
+                auto sym = MK::Symbol(arg->loc, cvar->name);
+                auto expr = MK::Send1(arg->loc, MK::Magic(arg->loc), core::Names::definedClassVar(),
+                                      location.copyWithZeroLength(), move(sym));
+                return make_node_with_expr<parser::Defined>(move(expr), location.join(arg->loc), move(arg));
+            }
+
+            // Desugar to `::Magic.defined_p(a, b, c)` or `::Magic.defined_p()`
+            ast::Send::ARGS_store args;
+            if (parser::NodeWithExpr::isa_node<parser::Const>(arg.get())) {
+                auto constPathNode = arg->takeDesugaredExpr();
+
+                if (auto currentConst = ast::cast_tree<ast::UnresolvedConstantLit>(constPathNode)) {
+                    while (!ast::isa_tree<ast::EmptyTree>(constPathNode)) {
+                        auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(constPathNode);
+                        if (lit == nullptr) {
+                            args.clear();
+                            break;
+                        }
+                        args.emplace_back(MK::String(lit->loc, lit->cnst));
+                        constPathNode = move(lit->scope);
+                    }
+
+                    absl::c_reverse(args);
+                }
```

Comment:
This test of `constPathNode` being an `UnresolvedConstantLit` is not necessary, as it's repeated inside the loop. As a bonus, removing it brings it even closer in line with the implementation being moved from PrismDesugar.cc.

```suggestion
                while (!ast::isa_tree<ast::EmptyTree>(constPathNode)) {
                    auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(constPathNode);
                    if (lit == nullptr) {
                        args.clear();
                        break;
                    }
                    args.emplace_back(MK::String(lit->loc, lit->cnst));
                    constPathNode = move(lit->scope);
                }

                absl::c_reverse(args);
```